orca-sql/src/main/kotlin/com/netflix/spinnaker/orca/sql/pipeline/persistence/SqlExecutionRepository.kt

```diff
@@ -308,6 +308,53 @@ class SqlExecutionRepository(
     return Observable.from(retrieveOrchestrationsForApplication(application, criteria, NATURAL_ASC))
   }
 
+  override fun retrieveExecutionsWithStatusInTimeWindow(
+    executionType: ExecutionType,
+    status: String,
+    updatedAtStart: Long,
+    updatedAtEnd: Long
+  ): List<Execution> {
+    return jooq.selectExecutions(
+      executionType,
+      conditions = {
+        it
+          .where(field("updated_at").greaterOrEqual(updatedAtStart))
+          .and(field("updated_at").lessOrEqual(updatedAtEnd))
+          .statusIn(listOf(ExecutionStatus.valueOf(status)))
+      },
+      seek = {
+        it
+          .orderBy(field("id").asc())
+      }
+    ).fetchExecutions().toMutableList()
+  }
+
+  override fun retrieveExecutionsWithSpecificStageTypesInTimeWindow(
+    executionType: ExecutionType,
+    status: String,
+    stageType: String,
+    updatedAtStart: Long,
+    updatedAtEnd: Long
+  ): List<Execution> {
+    // SELECT tP.*
+    //   FROM pipeline_stages AS tS
+    //   INNER JOIN pipelines AS tP
+    //     ON tS.execution_id = tP.id
+    //   WHERE JSON_EXTRACT(tS.body, "$.type") = "deploy" AND tS.updated_at > 0 AND tS.updated_at < 10000000000000 AND tS.status = "SUCCEEDED";
+
+    return jooq.select(
+      field("tP.id").`as`("id"), field("tP.body").`as`("body"))
```

Comment:
This almost makes me want to just write out the SQL with `?` parameters... let me know if there is a better way

Reply:
We _definitely_ do not want to use `JSON_EXTRACT` as part of a `WHERE` condition: This will cause a full table scan. You'll need to add a column for the stage type and index it.

Reply:
I am pretty sure `explain` locally was not showing a full table scan, but I will have to double check. It will have to parse a bunch of the statuses but it's going to be limited to the time window and the status clauses (both of which are required and with the index created on `updated_at` it should be pretty fast. but I will dig some more to make sure

Reply:
`JSON_EXTRACT` doesn't provide any magic: It requires inspection of a blob field, JSON deserialization and then filtering (probably via a temp table, now that I think of it, instead of a table scan). Try running that against the read replica in main and you'll have a bad time.